substrate/socket

```diff
@@ -98,6 +98,52 @@ namespace substrate
 	{
 		SUBSTRATE_API sockaddr_storage prepare(const socketType_t family, const char *const where,
 			const uint16_t port, const socketProtocol_t protocol = socketProtocol_t::tcp) noexcept;
+
+		inline int typeToFamily(const socketType_t type) noexcept
+			{
+				if (type == socketType_t::ipv4)
+					return AF_INET;
```

Comment:
We're surprised this compiles on CI as this macro and the other types this pulls over from the impl need these headers included to be defined:

```c++
#ifdef _WIN32
#	include <winsock2.h>
#else
#	include <sys/socket.h>
#	include <netdb.h>
#	include <arpa/inet.h>
#	include <netinet/in.h>
#	include <netinet/udp.h>
#endif
```

Given part of the idea was to reduce the visibility of this mess to the user and this would be an ABI break (requiring linking against the networking libraries of the OS for the user program rather than only substrate), perhaps it would be better to declare these helpers only in the header and keep them as full fat functions in socket.cxx - that way, the conversions are made available, but without this spill that we were originally trying to avoid?

To explain why that slew of headers are required:

* netinet/in.h is where sockaddr_in and sockaddr_in6 come from
* sys/socket.h is where the AF_, IPPROTO_ and SOCK_ macros come from
* arpa/inet.h is needed for sockadd_storage

The rest of the headers are dependencies for the later headers, as they don't include each other nicely - so `#include <netinet/in.h>` requires the previous 3 headers in order so compilation doesn't explode.

Reply:
> (requiring linking against the networking libraries of the OS for the user program rather than only substrate)

As I understand it, there's no leak of the networking libraries for consumers with this:

1. first of all, as you exposed pretty succintly in the last paragraph, the necessary headers are already provided by each OS's SDK
2. the needed socket libraries cannot be found with `dependency()`, they're found with `compiler.find_library` which does not provide include paths (this reinforces point 1 above)
3. such a leak will occur anyway, if substrate is built statically.

I'm not opposed to making these functions virtual, but I'm not sure if their complexity warrants four more entries in the export table.

Reply:
Thinking more clearly after talking about this with you, the leak would be in the external header now pulling in all the messy networking stuff from the OS in all cases.

Fair point with regards to linking - not sure what we were thinking earlier there. That said, given how the library is built, right now if you statically link substrate it does not "leak" this detail in our mind because the networking library only end up linked to the target binary via the transitive dependency, not an explicit requirement; meanwhile, for dynamic linking only substrate itself links them, with the program only linking substrate.

The bigger worry here is that particularly the Linux headers are incredibly messy and make a horrid number of symbols and macro defines in the global namespace, which has bitten us in the past and is the motivation behind containing that in impl/socket.cxx as much as possible.

For example, a macro is defined for every field in sockaddr_in and for every field in sockaddr_in6 to massage how you access the fields, and additional macros are then created for uint8_t, uint16_t and uint32_t access of some of the fields that are union'd in the structure.